odoo/websocket.py

```diff
@@ -570,66 +569,51 @@ def __exit__(self, *args):
         _ws_request_stack.pop()
 
     def _serve_websocket_message(self):
+        self._process_message()
         try:
-            self._process_message()
-        except json.JSONDecodeError:
-            raise InvalidWebsocketRequest('Invalid JSON data')
-        try:
-            self.registry = registry(self.db)
+            self.registry = Registry(self.db)
             self.registry.check_signaling()
         except (AttributeError, psycopg2.OperationalError, psycopg2.ProgrammingError) as exc:
             raise InvalidDatabaseException() from exc
 
         with closing(self.registry.cursor()) as cr:
             self.env = api.Environment(cr, self.session.uid, self.session.context)
-            endpoint, _ = self.ws_routing_adapter(self.registry._init_modules).match(self.path)
-            self._authenticate(endpoint)
-            return service_model.retrying(functools.partial(endpoint, **self.kwargs), self.env)
+            threading.current_thread().uid = self.env.uid
+            return service_model.retrying(self._serve_ir_websocket, self.env)
+
+    def _serve_ir_websocket(self):
+        ir_websocket = self.registry['ir.websocket']
+        rule, _ = ir_websocket._match(self.path)
+        ir_websocket._authenticate(rule.endpoint)
+        response = ir_websocket._dispatch(rule.endpoint)
+        return response
 
     def _process_message(self):
         """
             Extract required information from an incoming websocket
             message.
         """
-        self.message = json.loads(self.message)
+        try:
+            self.message = json.loads(self.message)
+        except json.JSONDecodeError:
+            raise InvalidWebsocketRequest('Invalid JSON')
         self.path = self.message.get('path')
-        self.kwargs = self.message.get('kwargs', {})
+        self.params = self.message.get('kwargs', {})
```

Comment:
See my comment above about parametric URLs and query-string. See also `HttpDispatcher.dispatch` and the `request.get_http_params` method.

IMO it should be

```py
        url = urllib.parse.url_parse(self.message.get('path', ''))
        self.path = url.path
        self.params = {
            **routing_args,  # should be passed as argument
            **urllib.parse.parse_qs(url.query),
            **self.message.get('kwargs', {}),
        }
```


```py